### PR TITLE
Resolve Item Loader bugs

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR fixes some convergence issues with the TokensLoader and add more tests to ensure we avoid regression.

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--19017.org.readthedocs.build/en/19017/

cc @carmocca @borda

### CODE REVIEW

File: tests/tests_data/streaming/test_dataset.py [hunk 1]
```diff
@@ -326,3 +329,204 @@ def test_try_create_cache_dir():
...
+
+    cache.done()
+    cache.merge()
+
+    for i in range(20):
+        sequence = cache[i]
+        assert sequence[0].item() == i * block_size
+        assert sequence[-1].item() == (i + 1) * block_size - 1
+
+    assert len(os.listdir(tmpdir)) == 6
```
Comment (awaelchli) [inline tests/tests_data/streaming/test_dataset.py:446]: 5 + 1 (index.json) file

**Review by awaelchli — APPROVED** [review]

File: src/lightning/data/streaming/dataset.py [hunk 2]
```diff
@@ -84,12 +83,15 @@ def __init__(
...
     def _create_cache(self, worker_env: _WorkerEnv) -> Cache:
         env = Environment(dist_env=self.distributed_env, worker_env=worker_env)
-        cache_path = _try_create_cache_dir(input_dir=self.input_dir.path, shard_rank=env.shard_rank)
-        cache_dir = copy.deepcopy(self.input_dir)
-        if cache_path:
-            cache_dir.path = cache_path
 
-        cache = Cache(input_dir=cache_dir, item_loader=self.item_loader, chunk_bytes=1, serializers=self.serializers)
+        if "this_" not in self.input_dir.path:
+            cache_path = _try_create_cache_dir(input_dir=self.input_dir.path, shard_rank=env.shard_rank)
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:89]: done !

File: src/lightning/data/streaming/item_loader.py [hunk 3]
```diff
@@ -149,6 +148,10 @@ def load_item_from_chunk(self, index: int, chunk_index: int, chunk_filepath: str
         if chunk_index not in self._mmaps:
             # TODO: Add deletion and memmap close
             chunk = self._chunks[chunk_index]
+
+            # Skip the header
+            # The number of items + the number of offsets (number of items in the chunk + 1)
+            # multiplied by their encoding dtype (np.uint32)
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning/data/streaming/item_loader.py:154]: added the comment.

File: tests/tests_data/streaming/test_dataset.py (hunk 1 above)
Comment (tchaton) [reply tests/tests_data/streaming/test_dataset.py]: done.

File: src/lightning/data/streaming/item_loader.py [hunk 4]
```diff
@@ -157,5 +160,5 @@ def load_item_from_chunk(self, index: int, chunk_index: int, chunk_filepath: str
         assert self._dtype
 
         buffer: bytes = self._buffers[chunk_index]
-        offset = self._dtype.itemsize * ((index - begin) if index >= begin else index + 1)
+        offset = self._dtype.itemsize * (index - begin) * self._block_size
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning/data/streaming/item_loader.py:163]: Done !

**Review by ethanwharris — APPROVED** [review]

File: .github/workflows/ci-tests-data.yml [hunk 5]
```diff
@@ -94,7 +94,7 @@ jobs:
         # NOTE: do not include coverage report here, see: https://github.com/nedbat/coveragepy/issues/1003
         run: |
           python -m coverage run --source lightning \
-            -m pytest -v --timeout=60 --durations=60
```
Comment (carmocca) [inline .github/workflows/ci-tests-data.yml:97]: You probably don't want to commit this. Or do you not care about all other tests?

Comment (tchaton) [reply .github/workflows/ci-tests-data.yml:97]: I am debugging a hanging test.

**Pull request merged**